genum/gen/traits.go

```diff
@@ -15,11 +32,103 @@ func (s TraitDescs) Less(i, j int) bool {
 	return s[i].Name < s[j].Name
 }
 
+func (s TraitDescs) getParsableUnderlying(u Underlying) TraitDescs {
+	out := make([]TraitDesc, 0, len(s))
+	for _, t := range s {
+		if t.Parsable && t.hasUnderlying(u) {
+			out = append(out, t)
+		}
+	}
+	return out
+}
+
+func (s TraitDescs) GetParsableUnderlyingString() TraitDescs {
+	return s.getParsableUnderlying(String)
+}
+
+func (s TraitDescs) GetParsableUnderlyingFloat64() TraitDescs {
+	return s.getParsableUnderlying(Float64)
+}
+
+func (s TraitDescs) GetParsableUnderlyingFloat32() TraitDescs {
+	return s.getParsableUnderlying(Float32)
+}
+
+func (s TraitDescs) GetParsableUnderlyingInt64() TraitDescs {
+	return s.getParsableUnderlying(Int64)
+}
+
+func (s TraitDescs) GetParsableUnderlyingUint64() TraitDescs {
+	return s.getParsableUnderlying(String)
+}
+
 // TraitDesc define a trait-- this is exposed for template use.
 type TraitDesc struct {
-	Name    string
-	TypeRef string
-	Traits  TraitInstances
+	Name     string
+	Type     types.Type
+	TypeRef  string
+	Parsable bool
+	Traits   TraitInstances
+}
+
+func (td *TraitDesc) extractUnderlying() (Underlying, bool) {
+	v, ok := td.Type.Underlying().(*types.Basic)
+	if !ok {
+		return Unknown, false
+	}
+	switch v.Kind() {
+	case
+		// types.UntypedInt,
```

Comment:
@drshriveer if I leave this in the generated code doesnt compile. I dont think want a match here

Reply:
This didn't make any sense to me, so I pulled down your branch uncommented, and verified that it compiled just fine for me?

Moreover, I discovered something else: with untypedInt and untypedString commented out the generated code looks like:

```go
	// first try and parse as a string
	*e, err = ParseEnumerableWithParsableTraits(value.Value)
	if err == nil {
		return nil
	}

 	// ints:
 	if sint64, err := strconv.ParseInt(value.Value, 10, 64); err != nil {
 		*e, err = ParseEnumerableWithParsableTraits(MyEnum(sint64))
 		if err == nil {
 			return nil
 		}
 	}
```

Where are the other traits? It should look more like:

```go
	// first try and parse as a string
	*e, err = ParseEnumerableWithParsableTraits(value.Value)
	if err == nil {
		return nil
	}
	

	// then try and parse for any string-like traits
	*e, err = ParseEnumerableWithParsableTraits(string(value.Value))
	if err == nil {
		return nil
	}
	
	// ints:
	if sint64, err := strconv.ParseInt(value.Value, 10, 64); err != nil {
		*e, err = ParseEnumerableWithParsableTraits(MyEnum(sint64))
		if err == nil {
			return nil
		}
		*e, err = ParseEnumerableWithParsableTraits(int(sint64))
		if err == nil {
			return nil
		}
	}
	
	// native parsing
	var v0 MyEnum
	if err := v0.UnmarshalYAML(value); err == nil {
		*e, err = ParseEnumerableWithParsableTraits(v0)
		if err == nil {
			return nil
		}
	}
```

Okay, better... Except!! I think even this is wrong... The way the enum is handled when it is an integer is kinda scary in the context of json/yaml/text unmarshaling.  I _think_ it really should be: 

```go
	// first try and parse as a string
	*e, err = ParseEnumerableWithParsableTraits(value.Value)
	if err == nil {
		return nil
	}
	
	// then try and parse for any string-like traits
	// ints:
	if sint64, err := strconv.ParseInt(value.Value, 10, 64); err != nil {
		*e, err = ParseEnumerableWithParsableTraits(int(sint64))
		if err == nil {
			return nil
		}
	}
	// native parsing
	var v0 MyEnum
	if err := v0.UnmarshalYAML(value); err == nil {
		*e, err = ParseEnumerableWithParsableTraits(v0)
		if err == nil {
			return nil
		}
	}
```

I'm to blame quite a bit here. There are some very subtle rules that I hadn't considered. 

1. any type that already implements the respective marshal interface shouldn't go through a casting layer. 
2. untyped strings are a special case that can be passed into the parsing function. Only typed strings need casting, and again, only if rule 1 allows it.